### ISSUE TITLE
Couldn't select pixels of TPFs in interact()

### ISSUE THREAD

**Issue by NaveenTime** [lightkurve/lightkurve#990]
**Problem description**
interact() is not working properly. I could not select the pixels in TPFs. When I try to select a pixel, the bottom row pixels automatically get selected. I could not select any pixel other than the bottom row pixels.

**Example**
```python
import lightkurve
tpf = lk.search_targetpixelfile("HL Tau", author='K2', campaign=13, cadence='long').download()
tpf.interact()
```

![Screenshot (354)](https://user-images.githubusercontent.com/65549947/109490657-8a697f80-7aae-11eb-9273-e8d9965ab510.jpg)
![Screenshot (356)](https://user-images.githubusercontent.com/65549947/109491164-3e6b0a80-7aaf-11eb-89b2-498b73482737.png)

**Expected behavior**
When I click a pixel in the tpf stamp,  the pixel actually should get selected. But it doesn't. Instead, the pixels of the lowest row get selected. I couldn't select any pixel other than the pixels of the lowest row.

**Environment**
-  Windows
-  lightkurve version 2.0.3
-  installation method - pip
-  Executed the code in Python 3 in JupyterNotebook (version 6.2.0)

**Comment by barentsen** [comment]
Oops, that sounds like a bug! :bug:

Thank you for taking the time to report this, @NaveenTime! :pray: 

@gully Would you be interested diagnosing the issue?  It may or may not be related to Lightkurve v2 changes, so a good first step may be to compare whether the bug can be reproduced in v1.11 vs v2.0.

**Comment by gully** [comment]
Hi @NaveenTime thanks for your interest in lightkurve especially my favorite tool `.interact()`! 🚀 

My first instinct was that it could be a Windows issue, since Windows may not have as-extensive user testing, so I did a fresh install of lightkurve on Windows 10 and was not able to reproduce this erroneous interact pixel-selection bug--- the pixels selected as intended.  

I seem to recall having some strange selection problems with dual monitors a while ago.  Are you using dual monitors?

**Comment by gully** [comment]
What browser are you using for Jupyter?  As another workaround and debugging step, I would recommend seeing if the bug reproduces in other browsers. I found no bug in Firefox with `bokeh` version `2.2.3`

**Comment by NaveenTime** [comment]
Hi @gully . I am not using dual monitors. I am using Jupyter Notebook in Google Chrome browser in Windows 10 with bokeh version 2.3.0.  I can reproduce the bug in Firefox also.

I checked for other tpf files and the bug still exists for any TPF file.

**Comment by gully** [comment]
Strange! My next hunch would be that this bug relates to javascript and the bokeh internals somehow?  We know that it works for similar conditions (Windows 10, Chrome, Jupyter, bokeh >=2.2).  Maybe the difference between `bokeh` 2.3 and 2.2.3 matters?  I doubt it, but that's one loose end, and the arena in which javascript interconnections are most accute.  Another weirdness I've seen in the past is that apps like `bokeh` that depend on javascript can find different javascript versions depending on if you have pointed to some locally or remotely, though that's a slightly esoteric problem and I doubt it's what's happening here.  I have also seen some weirdness depending on the mouse, and in particular the scroll wheel in a mouse can be hyperactive, again I doubt that's the issue but I'm brainstorming ways that have caused fickle behavior in the past in hopes it will cajole our brains into a solution.  

My top suggestion would be to downgrade to `bokeh` version 2.2.3 and see if the problem persists.

**Comment by NaveenTime** [comment]
I tried in Bokeh version 2.2.3. And it's working perfectly.  Also, note that when I used Bokeh version 2.3.0, the interact() didn't even show the default pipeline mask. In version 2.2.3, everything is working perfectly.

**Comment by gully** [comment]
Terrific! That's great news!  I am glad we have solved the problem.  Thank you again for using lightkurve and it's `.interact()` feature.

@barentsen we should probably:
- [ ] :heavy_minus_sign: make a patch to downgrade the bokeh version requirement to 2.2.3

I suspect this behavior would have been seen elsewhere, and I wonder if bokeh maintainers (cc @bryevdv or @mattpap) are aware of any change between bokeh 2.2.3 and 2.3 that would have affected python callbacks in such a way to alter selection behavior?

**Comment by barentsen** [comment]
Fantastic to hear we have a way to reproduce the issue, that's very helpful! :+1:

We'll want to have a look at the change log for bokeh 2.3.0 and try to understand why the behavior changed, and whether or not Lightkurve needs to be adapted.

I'm not a fan of changing the dependency to say `bokeh<2.3.0` for reasons [explained here](https://hynek.me/articles/semver-will-not-save-you/), but we could specify `bokeh != 2.3.0` if we were to ascertain that it is a temporary bug.

(I won't have time to dig into this myself, so any and all help welcome!)

**Comment by bryevdv** [comment]
I'm not aware of anything that should have changed with image selection. What would really help investigate is to isolate a minimal reproducer in a standalone script.

**Comment by gully** [comment]
Fascinating, I'm seeing a similar bug in Linux / Chrome /  bokeh=2.3.0 / and lightkurve=2.0.3.  I'm using TESScut TPFs.  The TPF mask selection defaults to the bottom row, and pixel selection is offset/fickle.  Now that I have it working on my machine I will attempt to get to the bottom of it and make a minimal reproducer in a standalone script.  Importantly there is now circumstantial evidence that this only arises in `bokeh=2.3.0`.  So if you are a lightkurve user seeing this issue, a workaround is to downgrade to `bokeh<2.3`

**Comment by bryevdv** [comment]
I can definitely believe a bug is possible, but the most helpful thing will be an MRE to investigate in isolation.

**Comment by bryevdv** [comment]
Was anyone ever able to check the JS console to see if there are any errors/warnings? cc @mattpap are there any `branch-2.3` changes that might intersect this?

**Comment by bryevdv** [comment]
@gully @NaveenTime @barentsen  FYI Bokeh 2.3.1 is released, it would be great if someone could confirm (or not) that it resolves the issue here

**Comment by barentsen** [comment]
@gully Time-permitting, would you be willing to see if you can still reproduce this issue with Bokeh 2.3.1?

**Comment by orionlee** [comment]
I happened to run into the same problem.

Environment:
- bokeh 2.3.1, lightkurve 2.0.9
- Windows platform, python 3.7.10

Furthermore, when I clicked the (incorrect) white pixels at the bottom, a python stacktrace showed up:

``` python
ERROR:bokeh.server.protocol_handler:error handling message
 message: Message 'PATCH-DOC' content: {'events': [{'kind': 'ModelChanged', 'model': {'id': '2225'}, 'attr': 'indices', 'new': [5]}], 'references': []} 
 error: ValueError("failed to validate Range1d(id='2257', ...).start: expected an element of either Float, Datetime or TimeDelta, got <Quantity -17.60428955 electron / s>")
Traceback (most recent call last):
  File "C:\pkg\_winNonPortables\Anaconda3\envs\lkv2\lib\site-packages\bokeh\server\protocol_handler.py", line 90, in handle
    work = await handler(message, connection)
```

**Comment by orionlee** [comment]
Update: after downgrading from bokeh 2.3.1 to 2.2.3,
- the aperture pixels are shown with white outline in initial plot correctly.
- when I try to select custom pixels by tapping them, the bokeh server error about receiving `Quantity` still shows up. So it might be a separate issue.

**Comment by bryevdv** [comment]
So lightkurve is doing looking to respond on changes to "indices":

https://github.com/lightkurve/lightkurve/blob/d9f9ba1145cad12ddc2b2bef0413827fc7e4a09c/src/lightkurve/interact.py#L860

But Bokeh set `image_indices` for at least two years, if not forever:

https://github.com/bokeh/bokeh/blame/branch-2.4/bokehjs/src/lib/models/glyphs/image_base.ts#L192

So speaking candidly don't understand how this could have ever worked at all.

**Comment by bryevdv** [comment]
If I had to guess tho, I might suggest looking in to this:

https://github.com/lightkurve/lightkurve/blob/d9f9ba1145cad12ddc2b2bef0413827fc7e4a09c/src/lightkurve/interact.py#L771-L776

Setting the selection indices in the middle of a callback on the selection indices seems like a very suspicious thing to do

**Comment by bryevdv** [comment]
OK I printed out the actual data in the callback and I believe this is the problem:
```
{'xx': array([[447., 448., 449., 450., 451., 452., 453., 454., 455., 456., 457.],
       [447., 448., 449., 450., 451., 452., 453., 454., 455., 456., 457.],
       [447., 448., 449., 450., 451., 452., 453., 454., 455., 456., 457.],
       [447., 448., 449., 450., 451., 452., 453., 454., 455., 456., 457.],
       [447., 448., 449., 450., 451., 452., 453., 454., 455., 456., 457.],
       [447., 448., 449., 450., 451., 452., 453., 454., 455., 456., 457.],
       [447., 448., 449., 450., 451., 452., 453., 454., 455., 456., 457.],
       [447., 448., 449., 450., 451., 452., 453., 454., 455., 456., 457.],
       [447., 448., 449., 450., 451., 452., 453., 454., 455., 456., 457.],
       [447., 448., 449., 450., 451., 452., 453., 454., 455., 456., 457.],
       [447., 448., 449., 450., 451., 452., 453., 454., 455., 456., 457.]])
```

This is not a 1d array. For all the "scatter" type glyphs (which is almost all glyphs. including `rect`) the CDS should contain only *one-dimensional* arrays. Perhaps something worked by accident previously, I'm not sure. But this has never been the expected data format. These arrays should be flattened.

**Comment by orionlee** [comment]
For the `ValueError` problem (due to `Quantity` passed to bokeh) when users tap to select custom aperture pixels with bokeh 2.2.3: it is a separate issue unrelated to bokeh 2.3.0/1. It is now tracked in #1033.

**Comment by keatonb** [comment]
FYI, I'm encountering this issue with lightkurve 2.0.9, bokeh 2.3.1, Mac OS X 10.14.5.  Downgrading to bokeh 2.2.3 fixed the issue. Note that I had to restart my Jupyter server after installing a different bokeh version before seeing the changes.

**Comment by orionlee** [comment]
FYI. the original issue still persisted with bokeh 2.3.2. 
In other words, pixel selection still only works with bokeh <= 2.2.3

**Comment by bryevdv** [comment]
@orionlee As I mentioned, there are problems above on the Lightkurve side that will have to be addressed:

> This is not a 1d array. For all the "scatter" type glyphs (which is almost all glyphs. including rect) the CDS should contain only one-dimensional arrays. Perhaps something "worked" by accident previously, I'm not sure. But this has never been the expected data format. These arrays should be flattened.

**Comment by gully** [comment]
Hey ya'll, yes Bryan is absolutely right, and the bug is clear--- I used 2D arrays for a `ColumnDataSource`, that only supports 1D arrays.  The fact that it worked at all is sort of a coincidence/miracle, and I thank Bryan for figuring that out, and also apologize for letting this issue fester and polluting your notifications.  Sorry Bryan!  We are grateful for all the awesome bokeh features that enabled this popular `interact()` dashboard.

The clear fix is to flatten and the arrays.  I started a PR (linked) in that direction, but it needs some spot-checking of array shapes that is tricky to do without some numpy surgery.  If any lightkurve users want to pair code with me for fun, tweet @ me: `gully_`  And yes to reiterate what previous folks have said, the temporary workaround is to use `bokeh <2.3`, but we emphasize that's temporary while we address the root cause.

**Comment by elnjensen** [comment]
I just encountered this bug myself on MacOS 10.15, python 3.9, lightkurve 2.0.9.   I first tried updating to the current codebase from github (lightkurve 2.0.10-dev as of June 2, 2021) before finding this thread, and I found that with that code, even downgrading bokeh to 2.2.3 doesn't fix the problem.  Maybe that's not surprising, but FYI for others encountering this. 

However, I *do* have it working with lightkurve 2.0.9 and bokeh 2.2.3.  @gully , I'd be happy to help with testing new code when you have a chance to work on this more.

**Comment by orionlee** [comment]
> and I found that with that code, even downgrading bokeh to 2.2.3 doesn't fix the problem.

@elnjensen Do you mean that you used the codes in PR #1057 or the codes in the latest `main` branch?
- Nothing is fixed in `main` branch yet
- For PR #1057, gully reported that it didn't quite work. However, it did work in my environment.

**Comment by elnjensen** [comment]
I mean that I tried using the `main` branch as it currently exists (using 'download ZIP'), but without that PR (which looks like it isn't yet merged).  So it would appear that some other code changes to `main` since 2.0.9 have caused bokeh 2.2.3 to no longer function as a workaround.  See attached screenshot, which shows that I get the bottom-row-selected behavior.  (I realize that just a screenshot doesn't show you *exactly* what lightkurve codebase I'm using, but you can see at least the version info after the first cell.) 

Given that there's WIP to fix this (and that it's maybe not clear why the function worked in the first place), I think this is a non-issue, but thought I'd report it. 

<img width="1366" alt="Screen Shot 2021-06-02 at 11 05 19 AM" src="https://user-images.githubusercontent.com/2545845/120511870-ce636d80-c398-11eb-943b-3b8b6849e158.png">

**Comment by orionlee** [comment]
@elnjensen On the issue with `main` with bokeh 2.2.3, that's good to know. It's a bit puzzling, as I use `main` with `bokeh 2.2.3` without any problem (on Windows, though the nature of the problem should not be dependent on OS).

It could be helpful if you do a quick test with the PR (download the zip  from [gully's PR branch](https://github.com/gully/lightkurve/tree/fix_for_Issue_990)). The PR as it is works for me (both bokeh 2.3+ and 2.2.3), but not for gully. It'd be helpful if there is a third-party to check if the fix works.

**Comment by gully** [comment]
I'm surprised the PR works at all, it was a first stab by doing a find-replace of the Column DataSource input shape.  I encourage folks to experiment with the PR to get the interact dashboard to work by seeing what changed in the PR so far and extrapolating to other lines of code that may need to adapt their shape.  The dashboard is tricky to maintain because the visual and interactive and Jupyter-specific output doesn't lend itself well to automated unit tests.

**Comment by davidjwilson** [comment]
Hi all, I just ran into the same issue. Downgrading to bokeh 2.2.3 (from 2.3.2) fixed it.

**Comment by yorkpf** [comment]
I have the same issue (apparently) 
<img width="1156" alt="Screen Shot 2021-07-31 at 2 15 29 pm" src="https://user-images.githubusercontent.com/76934896/127760168-616a2574-d9f9-4d01-8f7d-73525c86420c.png">
on a MacBook Pro, on which I am running the latest upgrade of MacOS Big Sur, version 11.15. I am using Anaconda Navigator 2.4 which in turn, runs Jupyter Notebook 6.3.0. Other packages of relevance are:lightkurve 2.0.10, bokeh 2.3.2, python 3.8.8.

Trouble is, as you can see, the image comes up without any boxes “chosen” (by a white border), so it is far from clear which pixels are being used to generate the light curve. What is worse is that no amount or combination of mouse clicking or shift-clicking seems able to select or deselect even a single pixel in the right-hand diagram. Occasionally, I also get the situation where the lowest horizontal layer of boxes is "selected", as reported by others.

I would welcome a solution because I am getting used to operating at the Notebook level. I just do not have the knowledge to troubleshoot the underlying packages ... Paul

<img width="1156" alt="Screen Shot 2021-07-31 at 2 15 29 pm" src="https://user-images.githubusercontent.com/76934896/127760235-92c01f99-7cee-4701-8e98-2fff8ed4d56c.png">

**Comment by orionlee** [comment]
@yorkpf The workaround is to downgrade bokeh from v2.3.x to 2.2.3.

**Comment by yorkpf** [comment]
Thanks for that Sam. Would this downgrade be likely to bring a whole lot of (complicated) dependencies with other packages into play?

**Comment by yorkpf** [comment]
Sounds straightforward (downgrade bokeh to 2.2.3). When I try this via Anaconda, it fails after more than 30 minutes with an "UnsatisfiableError". Using the Terminal command ends up in much the same way ... Is there some more straightforward way to downgrade a package without compromising the existing environment?

**Comment by orionlee** [comment]
Downgrading to bokeh 2.2.3 does not cause any dependency issues in my experience and others who have reported in this issue.

On the "UnsatisfiableError" during downgrade, I have not seen it with bokeh (Anaconda on Windows 10, python 3.7.10), I'd suggest you try some of the following:
1. if you're sure you don't need bokeh 2.3.x (lightkurve does not require it), bypass Anaconda by using `pip` to downgrade, or
2. create a new environment, install `lightkurve`, downgrade bokeh to 2.2.3 (then install other packages if you require any extra), or
3. "UnsatisfiableError" should list out the actual incompatibility, inspect them and see if you can resolve it.

**Comment by yorkpf** [comment]
I took the path of going via pip to uninstall the current bokeh and then install bokeh 2.2.3. After re-installing lightkurv, the notebook worked OK. Thanks.

**Comment by orionlee** [comment]
#### Tips for the workaround: downgrade to bokeh 2.2.3 

With the latest lightkurve (v.2.1.1) and its dependencies, downgrading to bokeh 2.2.3 (from the default 2.4.2 as of the time of this post) could be a bit tricky:
- bokeh 2.2.3 is incompatible with the installed jinja2 v 3.x
- downgrade from 2.4.2 to 2.2.3 might lead to a corrupted bokeh (this might be specific to anaconda)

The following is what I do to get the downgrade work with anaconda:
- `conda remove bokeh --force`
- forcefully remove any lingering `bokeh`
  - Go to Anaconda environment package directory, `<Anaconda Home>\envs\<env name>\Lib\site-packages`
  - remove directories `bokeh` and `bokeh-<version number>.dist-info` if they are there
- install bokeh 2.2.3 along with a compatible jinja2 `conda install bokeh=2.2.3 jinja2\<3`  

---

You can use the following script as a sanity test the installation
```
python -c "import bokeh; print(bokeh.__path__); print('bokeh: ', bokeh.__version__); from bokeh.io import output_notebook;"
```
- it should show bokeh 2.2.3 is installed without error
- if the `from bokeh.io import output_notebook;` fails, it is probably because you have a corrupted bokeh (a mix of 2.2.3 and 2.4.2 are installed)




```

**Comment by barentsen** [comment]
Update: a fix for this issue has been included in the Lightkurve v2.2.0 release.
